PR: Improve UX of the Plots plugin

ccordoba12 opened the pull request:
## Description of Changes

- Move `Fit plots` action to the plugin toolbar and make plots be fitted automatically to the pane the first time they are generated. That makes the zoom in/out buttons be enabled by default, which is a much better usability experience than having them disabled.
- Disabling `Fit plots` shows them at full size.
- Save the `Fit plots` and zoom level for every plot. That way, we'll be able to show each plot in the specific way users want to see them.
- Implement drag and drop of thumbnails so that users can manually reorganize plots as they see fit. 
- Improve organization of buttons in its main toolbar and increase default width of thumbnail bar to go with this change.
- Make doing double click on a plot show it at full size.

### Visual changes

* Main toolbar
    - Before 
        ![imagen](https://github.com/spyder-ide/spyder/assets/365293/19f9e8c1-52fa-4205-8695-7581c867e0d4)
    
    - After
        ![imagen](https://github.com/spyder-ide/spyder/assets/365293/a6f4ee35-6cd7-4ac2-8d24-fb608021436e)

* `Fit plots` button

    ![fit-button](https://github.com/spyder-ide/spyder/assets/365293/ef65e638-1dab-4bb0-90af-0da1f7714bc6)

* Separate zoom level for each plot

    ![zoom-per-plot](https://github.com/spyder-ide/spyder/assets/365293/9c8eb647-2290-4a74-a65a-33a53a9b640d)

* Drag and drop

    ![plot-drag-and-drop](https://github.com/spyder-ide/spyder/assets/365293/c5574a6c-a307-4011-84be-a145a443773d)

* Zoom in on double click

    ![plot-double-click](https://github.com/spyder-ide/spyder/assets/365293/ff847e1a-1184-4ae0-b560-5b783f6dadc2)

### Issue(s) Resolved

Fixes #21899.
Fixes #12158
Part of https://github.com/spyder-ide/ux-improvements/issues/19

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct:

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
@dalthviz, could you review if functionality of these changes is working as expected? Thanks!

About the UI/UX changes, I already discussed with the UX team and they agreed with them.

dalthviz commented:
Gave this a check and seems like the double click is changing the zoom factor value even when the option to `Fit plots to window` is enabled causing a sudden zoom factor change when the option gets disabled again. Probably the zoom factor when disabling the `Fit plots to window` should be the one the plot have before the option was enabled but not sure if that is intended. Here a preview of that behavior:

![plots_double_click](https://github.com/spyder-ide/spyder/assets/16781833/17de851c-9fd9-495f-b98f-452076e5f872)

Other than that seems like things are working :+1:

ccordoba12 commented:
Great catch @dalthviz! I disabled doing zooming in and out with double clicks when `Fit plots to window` is enabled. That actually doesn't make much sense because the zoom factor is fixed to the size of the Plots pane.

dalthviz commented:
Gave another check and although the click with the option enabled is fixed, seems like now (or at least I didn't noticed that before 🤔) a double double click needs to be done to zoom up to 100%:

![plots_double_double_click](https://github.com/spyder-ide/spyder/assets/16781833/a77d315b-1974-465c-89da-61eb30f85ca1)

I think is related with passing to another plot when the previous one was left at 100% 🤔

pep8speaks commented:
Hello @ccordoba12! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









* In the file [`spyder/utils/icon_manager.py`](https://github.com/spyder-ide/spyder/blob/1d9bcf45be1d1d1ffc1ce9a96d4aedefc1fc7fe4/spyder/utils/icon_manager.py):

> [Line 208:80](https://github.com/spyder-ide/spyder/blob/1d9bcf45be1d1d1ffc1ce9a96d4aedefc1fc7fe4/spyder/utils/icon_manager.py#L208): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)

ccordoba12 commented:
Thanks for the additional review @dalthviz! After considering it carefully, I decided to push the following improvements to make the UX more intuitive:

* Move the fit action to the main toolbar as a toggle and check it after every a new plot is generated.
* When that action is unchecked the plot is shown at full size.
* When the user does zoom in/out on a plot, that action is automatically unchecked.
* Preserve the zoom level of each plot so that switching plots in the thumbnail bar shows them exactly as users were seeing them the last time. 

I updated the OP with these new changes accordingly.

dalthviz commented:
Gave this another check and seems like now, besides the previous behavior (double double click need to effectivly change between zoom factors when changing between plots), there is also a possiblitity to make the zoom factor to get stuck in 100% 🤔:

![double_double_click_zoom_2](https://github.com/spyder-ide/spyder/assets/16781833/fdf5983c-4137-49e4-849e-cf4dd1538e90)

ccordoba12 commented:
@dalthviz, I didn't understand that problem but I do now. So, I rewrote one of my previous commits to fix it. That removed the ability to zoom out to the previous size with a second double-click. But I think that's not so important now that I added the auto-fit button to the toolbar.

So, please check again.

dalthviz commented:
Checked again and now I'm unable to reproduce the mentioned behaviors so seems like things are working as described :+1: The only suggestion that came to mind while checking the new behavior is to maybe add a shortcut for the new auto-fit button.

dalthviz commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -423,13 +454,42 @@ def show_context_menu(self, qpoint):
+            self.current_thumbnail.scalefactor = self.scalefactor
+
+        # Restore scrollbar values
+        QTimer.singleShot(
+            20,
+            lambda: self.verticalScrollBar().setValue(
+                self.current_thumbnail.vscrollbar_value
+            ),
+        )
+        QTimer.singleShot(
+            20,
+            lambda: self.horizontalScrollBar().setValue(
+                self.current_thumbnail.hscrollbar_value
+            ),
+        )
```
Maybe an explanation of why these timers need to be set here could be worthy

ccordoba12 replied:
Done at the top of these timers.

ccordoba12 commented:
> The only suggestion that came to mind while checking the new behavior is to maybe add a shortcut for the new auto-fit button.

Good idea. I set `Ctrl+0` for this action.

dalthviz reviewed (approved):
Thanks @ccordoba12 ! Things LGTM :+1: Maybe adding some sort of test could be nice to have but leaving this approved so feel free to merge

ccordoba12 commented:
I added a couple of tests to cover the new auto-fit and zoom level per plot functionality, which I think are the most important ones. So, this is ready now.
